### PR TITLE
feat(freetype): simplify font create routines

### PR DESCRIPTION
**Pull request by W-Mai**
Help us review this PR! Anyone can [approve it or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### Description of the feature or fix

Simplify freetype font create routines and some bug fix

### Checkpoints
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html)

### CODE REVIEW

**Comment by kisvegabor** [comment]
Could you also take care of [this](https://github.com/lvgl/lvgl/pull/5196#issuecomment-1879684691)? It affects freetpye, tiny_ttf, and binfont_loader. 

I've just noticed that the `cache_size` parameter of tiny_ttf is not used so [the `..._ex()` functions](https://github.com/lvgl/lvgl/blob/master/src/libs/tiny_ttf/lv_tiny_ttf.h#L37) can be removed.

**Comment by W-Mai** [comment]
> Could you also take care of [this](https://github.com/lvgl/lvgl/pull/5196#issuecomment-1879684691)? It affects freetpye, tiny_ttf, and binfont_loader.

Oh, yes.

1. [ ] new format attr in `lv_font_glyph_dsc_t` to replace `bpp`
2. [ ] new API form to create font

> I've just noticed that the `cache_size` parameter of tiny_ttf is not used so [the `..._ex()` functions](https://github.com/lvgl/lvgl/blob/master/src/libs/tiny_ttf/lv_tiny_ttf.h#L37) can be removed.

In fact, this is my mistake. At that time, I hardcoded a 128, which should be configurable. It was my mistake.

**Comment by kisvegabor** [comment]
> In fact, this is my mistake. At that time, I hardcoded a 128, which should be configurable. It was my mistake.

128 means 128 kB? If it's for the whole TinyTTF module it should be an `lv_conf.h` config.  However, if I recall correctly we were talking about using a single unified font cache for tiny_ttf and freetype. Was it changed or is it still in progress? Having separate caches makes sens too.

**Comment by W-Mai** [comment]
> 128 means 128 kB? If it's for the whole TinyTTF module it should be an `lv_conf.h` config. 
However, if I recall correctly we were talking about using a single unified font cache for tiny_ttf and freetype. Was it changed or is it still in progress? Having separate caches makes sens too.

128 now means that the upper limit is 128 glyph. At present, my intention is still to merge freetype and tiny_ttf, but it seems to be a little complicated.

So at present, I have slightly lowered this priority🙃.

File: src/libs/freetype/lv_freetype.c [hunk 1]
```diff
@@ -249,23 +256,51 @@ const char * lv_freetype_get_pathname(FTC_FaceID face_id)
...
     if(dsc->render_mode == LV_FREETYPE_FONT_RENDER_MODE_BITMAP) {
-        return lv_freetype_image_font_create(dsc);
+        draw_data_cache = lv_freetype_create_draw_data_image();
     }
     else if(dsc->render_mode == LV_FREETYPE_FONT_RENDER_MODE_OUTLINE) {
-        return lv_freetype_outline_font_create(dsc);
+        draw_data_cache = lv_freetype_create_draw_data_outline();
     }
     else {
         LV_LOG_ERROR("unknown render mode");
```
Comment (FASTSHIFT) [inline src/libs/freetype/lv_freetype.c:280]: LV_LOG_ERROR("unknown render mode: %d", dsc->render_mode);

File: src/libs/freetype/lv_freetype.c [hunk 2]
```diff
@@ -249,23 +256,51 @@ const char * lv_freetype_get_pathname(FTC_FaceID face_id)
...
+
+    dsc->cache_node->draw_data_cache = draw_data_cache;
+
+    return true;
 }
 
-/**********************
- *   STATIC FUNCTIONS
- **********************/
+static void freetype_on_font_set_cbs(lv_freetype_font_dsc_t * dsc)
```
Comment (FASTSHIFT) [inline src/libs/freetype/lv_freetype.c:294]: freetype_on_font_set_ops

File: src/libs/freetype/lv_freetype.c [hunk 3]
```diff
@@ -249,23 +256,51 @@ const char * lv_freetype_get_pathname(FTC_FaceID face_id)
...
+{
+    lv_freetype_set_cbs_glyph(dsc);
+    if(dsc->render_mode == LV_FREETYPE_FONT_RENDER_MODE_BITMAP) {
+        lv_freetype_set_cbs_image_font(dsc);
+    }
+    else if(dsc->render_mode == LV_FREETYPE_FONT_RENDER_MODE_OUTLINE) {
+        lv_freetype_set_cbs_outline_font(dsc);
```
Comment (FASTSHIFT) [inline src/libs/freetype/lv_freetype.c:301]: else

File: src/libs/freetype/lv_freetype.c (hunk 2 above)
Comment (W-Mai) [reply src/libs/freetype/lv_freetype.c:294]: Great!

File: src/libs/freetype/lv_freetype.c (hunk 3 above)
Comment (W-Mai) [reply src/libs/freetype/lv_freetype.c:301]: Great!

File: src/libs/freetype/lv_freetype.c (hunk 1 above)
Comment (W-Mai) [reply src/libs/freetype/lv_freetype.c:280]: Great advice

**Comment by kisvegabor** [comment]
> 128 now means that the upper limit is 128 glyph. At present, my intention is still to merge freetype and tiny_ttf, but it seems to be a little complicated.

Shouldn't it be size based?

**Review by FASTSHIFT — APPROVED** [review]

**Comment by W-Mai** [comment]
> merge freetype and tiny_ttf, but it seems to be a little complicated.

Now freetype uses count, which is quite convenient to use. To use a specific memory size or a number, I have no ideas about which one is better.

**Comment by kisvegabor** [comment]
Character count based could mean small and large memory usage based on font size. So I vote for memory based is possible.

**Pull request merged**